src/mpyl/steps/deploy/kubernetes.py

```diff
@@ -69,26 +76,44 @@ def get_endpoint(builder: ChartBuilder) -> str:
     def execute(self, step_input: Input) -> Output:
         builder = ChartBuilder(step_input)
         chart = to_service_chart(builder)
+        project = step_input.project_execution.project
 
-        deploy_result = deploy_helm_chart(
-            self._logger, chart, step_input, builder.release_name
+        chart_path = write_helm_chart(
+            self._logger,
+            chart,
+            project.target_path,
+            step_input.run_properties,
+            builder.release_name,
         )
-        if deploy_result.success:
-            hostname = self.try_extract_hostname(chart, builder.project.name)
-            url = None
-            if hostname:
-                self._logger.info(
-                    f"Service {step_input.project_execution.name} reachable at: {hostname}"
-                )
-                endpoint = self.get_endpoint(builder)
-                url = f"{hostname}{endpoint}"
-            artifact = input_to_artifact(
-                ArtifactType.DEPLOYED_HELM_APP,
-                step_input,
-                spec=DeployedHelmAppSpec(url=f"{url}"),
+
+        if step_input.install:
+            deploy_result = deploy_helm_chart(
+                self._logger, chart_path, step_input, builder.release_name
             )
-            return Output(
-                success=True, message=deploy_result.message, produced_artifact=artifact
+
+            if deploy_result.success:
+                hostname = self.try_extract_hostname(chart, project.name)
+                url = None
+
+                if hostname:
+                    self._logger.info(
+                        f"Service {step_input.project_execution.name} reachable at: {hostname}"
+                    )
+                    url = f"{hostname}{self.get_endpoint(builder)}"
+                deploy_result.produced_artifact = input_to_artifact(
```

Comment:
This doesn't feel very right 🫤 

My first instinct was to ask to move this whole block to `deploy_helm_chart()`, but then I noticed we also call this method from `kubernetes_job.py` without doing this. In that case, I feel like we should stop creating the `Output` inside of `deploy_helm_chart()` and always do it here, to prevent having to mess with this very object after creating (immutability is good!).

Let's discuss tomorrow.

Reply:
Yeah so this is precisely the part that I mentioned yesterday about not being too happy with the refactor. The point is that `deploy_helm_chart()` doesn't explicitly create an `Output`. It calls `custom_check_output` in the end which always returns a generic `Output` by design. What we do here is overwriting it to a more specific Output for this function.

Reply:
I see. I think it would be clear to create a new `Output` at this point instead of modifying the original one: 
```python
deploy_result = deploy_helm_chart(…)

if deploy_result.success:
    …
    
    deploy_result = Output(
        success=True,
        message=f"Helm charts written to {chart_path}",
        produced_artifact=input_to_artifact(
            ArtifactType.DEPLOYED_HELM_APP,
            step_input,
            spec=DeployedHelmAppSpec(url=f"{url}"),
        )
```

And if the only thing we need from `deploy_helm_chart()` is to know whether it succeeded or not, we can skip the `Output` altogether and simply return a boolean?
```python
if deploy_helm_chart(…):
    …
    
    deploy_result = Output(
        success=True,
        message=f"Helm charts written to {chart_path}",
        produced_artifact=input_to_artifact(
            ArtifactType.DEPLOYED_HELM_APP,
            step_input,
            spec=DeployedHelmAppSpec(url=f"{url}"),
        )
```
With some luck this is the only place in the project where we modify `Output` and we can decorate it with:
```python
@dataclass(frozen=True)
```

Reply:
> if the only thing we need from deploy_helm_chart() is to know whether it succeeded or not

No, if it fails we want to know the error message.

Anyway, I _tried_ to make `Output` frozen but I don't think it's possible because we use it as a `@yaml_object`. I don't fully understand how that process works, but it can't handle frozen dataclasses (the unit tests saved me here). None of the classes in mpyl with `@yaml_object` are frozen in fact.

I did ensure we no longer do updated to Output objects in https://github.com/Vandebron/gh-mpyl/pull/16/commits/9d39878a8f92d678c4ad915a7dd0be481e0e03a4

Reply:
I'm also not super familiar with how that works so we can definitely tackle that in a follow-up PR! 💪